### DISCUSSION TITLE
Async ItemsSource with SelectedItem/Value in ComboBox

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#13595] [Q&A]
Background:

I'm binding a ComboBox SelectedValue/SelectedItem and ItemsSource to a view model. However, the ItemsSource is async set only on demand.

Problem:

As mentioned, ItemsSource is asynchronously populated from a background service. This can take some time to fetch and filter all the appropriate values so it isn't done right away in the view model. However, the SelectedValue/Item has no reason for any delay and is set immediately when restoring from previously saved values. The issue is, because the ItemsSource collection doesn't yet exist, the SelectedValue/Item is automatically cleared by the binding to the ComboBox because the selected value doesn't exist in the collection (which again doesn't exist quite yet). Clearing the SelectedValue/Item because it doesn't exist in a non-existent ItemsSource is a problem if the user has previously already made a selection.

Question(s):

 1. Does anyone have ideas how to stop the ComboBox from clearing the selected value when the ItemsSource doesn't yet exist. Should this logic be added to the control itself perhaps?
 2. There is a rather ugly work-around in place in the view model at the moment where the property setter is just ignored in some cases. Anyone have better ideas for a work-around here?
 3. Is this a situation anyone else has run across, if so how did you solve it?

**Comment by timunie** [comment]
> Does anyone have ideas how to stop the ComboBox from clearing the selected value when the ItemsSource doesn't yet exist. Should this logic be added to the control itself perhaps?

In the ViewModel you could cancel this in the setter. But this is probably only a workaround for now. Also you may want to implement your custom SelectionModel to have better control over the selection itself. 

> Is this a situation anyone else has run across, if so how did you solve it?

I remember having similar issues in WPF when bound to selectedItem. SelectedValue worked iirc.

**Reply by robloo** [reply]
>> Does anyone have ideas how to stop the ComboBox from clearing the selected value when the ItemsSource doesn't yet exist. Should this logic be added to the control itself perhaps?
>
> In the ViewModel you could cancel this in the setter. But this is probably only a workaround for now. Also you may want to implement your custom SelectionModel to have better control over the selection itself.

Yes, that is the work-around already in place. There is an `isItemsSourceLoaded` property that carefully tracks state elsewhere in the view model. Only when it is true is the setter allowed to change anything.

> I remember having similar issues in WPF when bound to selectedItem. SelectedValue worked iirc.

Current code is using Avalonia's SelectedValue but testing SelectedItem behaves the same. If there really is a difference from WPF it might be good to investigate further. However, that isn't a priority and I probably won't spend the time to check myself.

---

I was curious if anyone had other ideas on how to address this. If not I'll mark this as the answer and close it in a while. Thanks for the ideas.

**Comment by geferon** [comment]
Did you manage to find a solution or workaround to this that wasn't modifying the VM? I'm currently working with a set of DTO's that are shared between multiple projects, Avalonia only being one of them, and I can't really implement it like so.

**Comment by MariusJaeger** [comment]
Create itemsSource wuth the selected value. Then later the async task can replace itemsSource with the List of items.
